### DISCUSSION TITLE
Fast Data | SVC | "customFx is not a function" error

### DISCUSSION THREAD

**Discussion by sfontana9** [mia-platform/community#438] [Q&A] [answered]
### What product are you having troubles with?

Fast Data

### What Console version are you using?

12.0.2

### Description

Hello, while aggregating from a message coming from the Single View Trigger Generator, my Single View Creator is throwing this error

`{"level":50,"time":1711120603438,"pid":7,"hostname":"xxxx-svc-low-code-676bffb666-dps9q","err":{"type":"TypeError","**message":"customFx is not a function",**"stack":"TypeError: customFx is not a function\n    at Mapper.getResolvedValueForField (/home/node/app/node_modules/@mia-platform-internal/fast-data-automation-lib/lib/aggregation/Mapper.js:67:33)\n    at Mapper.resolve (/home/node/app/node_modules/@mia-platform-internal/fast-data-automation-lib/lib/aggregation/Mapper.js:133:64)\n    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at async /home/node/app/getFunctions.js:139:32\n    at async executeAggregation (/home/node/app/node_modules/@mia-platform-internal/single-view-creator-lib/lib/single-view-creation.js:245:29)\n    at async KafkaMessagesHandler.manageAggregation (/home/node/app/node_modules/@mia-platform-internal/single-view-creator-lib/lib/KafkaMessagesHandler.js:243:7)\n    at async KafkaMessagesHandler.handleKafkaMessage (/home/node/app/node_modules/@mia-platform-internal/single-view-creator-lib/lib/KafkaMessagesHandler.js:224:7)\n    at async Runner.eachBatch (/home/node/app/node_modules/@mia-platform-internal/single-view-creator-lib/lib/KafkaMessagesHandler.js:144:15)\n    at async Runner.processEachBatch (/home/node/app/node_modules/kafkajs/src/consumer/runner.js:274:7)\n    at async onBatch (/home/node/app/node_modules/kafkajs/src/consumer/runner.js:449:9)"},"msg":"error while handling Projection Changes kafka message"}`

I can understand this is something related to custom functions in aggregation file, but i can't manage to find the problem. 

### Actual Outcome

The aggregation fails

### Expected Outcome

The aggregation succeeds

**Comment by flower-of-the-bridges** [comment]
Hi @sfontana9 !

Is there a reference to a `customFx` variable either in `aggregation.json` or in any custom function that is mounted in your single view creator service?

**Reply by sfontana9** [reply]
Hi, no, i searched the whole Configurations of my project on GitLab and the only reference i can find is inside the readme.MD at fast-data-files/low-code.

```
singleViews.json: file that will contain information about the microservices where each single view configuration created will be placed. More information about this file are shown in the [next section](https://github.com/mia-platform/community/discussions/438#compiling-the-configurations);

/configurations: this folder contains the Fast Data Low Code configurations

/{singleView}: contains the configuration files for a particular single view, such as

singleViewKey.ts: contains the identifier schema;

aggregation.ts: contains the aggregation schema;

erSchema.ts: contains the ER schema

{customFx.js}: additional custom function files (with .js extension);

validator.js: function which contains assertion about the single view.

utils: this folder contains util functions with .js extension that can be shared among different single view. To import an util function into a custom function, you have to interpolate the env variable of the relative path, that will be discarded when the function will be compiled:
```

**Reply by flower-of-the-bridges** [reply, marked as answer]
Hi @sfontana9 

You can ignore `fast-data-files` folders since its no longer used by console (if you're having a console higher than v12.0.0)

The `customFx` mentioned in the logs refers to a custom function directive in your `aggregation.json` file, i.e. a field that has a mapping equal to `__fromFile__[<your fx name>.js]`

Inside this file (that can be managed by the advanced mode of the single view aggregation tab), you need to export **exactly** a function:

```js
module.exports = async function(logger, mongoClient, dependencies) {
  // your logic
}
```
